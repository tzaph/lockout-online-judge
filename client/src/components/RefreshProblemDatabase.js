import React, { useEffect, useState } from 'react'
import { Alert } from 'react-bootstrap';
import { getDatabase, ref, set, update } from "firebase/database";
import { Link, useNavigate } from 'react-router-dom';
import axios from "axios"

export default function RefreshProblemDatabase() {
  const [error, setError] = useState('');
  const [message, setMessage] = useState("Loading...");
  const navigate = useNavigate();

  useEffect(() => {
    for (let ratingValue = 800; ratingValue <= 3600; ratingValue += 100) {
      setTimeout(() => {
        if (ratingValue <= 3500) {
          setMessage("Loading problems of rating " + ratingValue)
          axios.get(`https://codeforces.com/api/problemset.problems?tags=`)
              .then(res => {
                let problemsetList = res.data.result.problems
                problemsetList = problemsetList.filter(
                  (problem) => 
                  problem.rating == ratingValue
                )

                const db = getDatabase()
                set(ref(db, 'problems/' + ratingValue), {
                  problemsetList: problemsetList
                });
              }).catch (err => {
                setError(err)
              })
        } else {
          setMessage("Done!")
        }
      }, 30 * (ratingValue - 700))
    }
  }, [])

  return (
    <div className="auth-form-container">
      <h2>Refreshing Codeforces Problem Database</h2>
      {error && <Alert variant='danger'>{error}</Alert>}
      {message == "Done!" ? <div className="link-button"><Link to="/">{message}</Link></div> : <p>{message}</p>}
    </div>
  );
}
