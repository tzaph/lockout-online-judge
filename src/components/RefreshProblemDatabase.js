import React, { useEffect, useState } from 'react'
import { Alert } from 'react-bootstrap';
import { getDatabase, ref, set, update } from "firebase/database";
import { Link, useNavigate } from 'react-router-dom';
import axios from "axios"

export default function RefreshProblemDatabase() {
  const [error, setError] = useState('');
  const [message, setMessage] = useState("Loading...");
  const [loading, setLoading] = useState(false);
  const navigate = useNavigate();

  const ratingValues = [
    800, 900, 1000, 1100, 1200,
    1300, 1400, 1500, 1600, 1700,
    1800, 1900, 2000, 2100, 2200,
    2300, 2400, 2500, 2600, 2700,
    2800, 2900, 3000, 3100, 3200,
    3300, 3400, 3500
  ]

  const refreshProblems = async() => {
    try {
      console.log("hi");
      const db = getDatabase()
      ratingValues.map(ratingValue => {
        set(ref(db, 'problems/' + ratingValue), {
          problemsetList: []
        });
      })

      const apiCall = await axios.get(
        `https://codeforces.com/api/problemset.problems?tags=`
      );
      ratingValues.map(ratingValue => {
        console.log(ratingValue);
        let problemsetList = apiCall.data.result.problems;
        problemsetList.filter(
          (problem) => 
          problem.rating >= ratingValue && problem.rating <= ratingValue
        );
        console.log(problemsetList.length);

        const db = getDatabase()
        update(ref(db, 'problems/' + ratingValue), {
          problemsetList: problemsetList
        });
      })

      setLoading(true);
      setError('');
      navigate('/');
      navigate(0);
    } catch (err) {
      setError(`Failed to refresh problems, ${err}`);
    }
    setLoading(false);
  }

  useEffect(() => {
    const apiCall = axios.get(
      `https://codeforces.com/api/problemset.problems?tags=`
    );

    for (let ratingValue = 800; ratingValue <= 3600; ratingValue += 100) {
      setTimeout(() => {
        if (ratingValue <= 3500) {
          setMessage(ratingValue)
          axios.get(`https://codeforces.com/api/problemset.problems?tags=`)
              .then(res => {
                console.log(ratingValue);
                let problemsetList = res.data.result.problems
                problemsetList = problemsetList.filter(
                  (problem) => 
                  problem.rating >= ratingValue && problem.rating <= ratingValue
                )
                problemsetList.map((problem) => {
                  console.log(problem.rating);
                })

                const db = getDatabase()
                set(ref(db, 'problems/' + ratingValue), {
                  problemsetList: problemsetList
                });
              }).catch (err => {
                setError(err)
              })
        } else {
          setMessage("Done!")
        }
      }, 30 * (ratingValue - 700))
    }
  }, [])

  return (
    <div className="auth-form-container">
      <h2>Refreshing Codeforces Problem Database</h2>
      {error && <Alert variant='danger'>{error}</Alert>}
      <div className="link-button">
        <Link to="/">{message}</Link>
      </div>
    </div>
  );
}
