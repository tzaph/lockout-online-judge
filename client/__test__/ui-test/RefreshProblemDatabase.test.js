import React from "react";
import renderer from "react-test-renderer";
import RefreshProblemDatabase from "../../src/components/RefreshProblemDatabase";
import { BrowserRouter as Router } from 'react-router-dom';

describe("<RefreshProblemDatabase />", () => {
  it("Renders correctly", () => {
    const tree = renderer.create(<Router><RefreshProblemDatabase /></Router>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
