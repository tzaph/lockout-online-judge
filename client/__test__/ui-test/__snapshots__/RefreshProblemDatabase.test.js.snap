// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RefreshProblemDatabase /> Renders correctly 1`] = `
<div
  className="auth-form-container"
>
  <h2>
    Refreshing Codeforces Problem Database
  </h2>
  <p>
    Loading...
  </p>
</div>
`;
